data_structures/binary_trees/segment_tree.c

```diff
@@ -0,0 +1,231 @@
+/**
+ * @file segment_tree.c
+ * @brief segment trees with only point updates
+ * @details
+ * This code implements segment trees. Segment trees are general structures
+ * which allow range based queries in a given array in logN time.
+ * Segment tree with point updates allow update of single element in the array
+ * in logN time.
+ * [Learn more about segment trees
+ * here](https://codeforces.com/blog/entry/18051)
+ * @author [Lakhan Nad](https://github.com/Lakhan-Nad)
+ */
+
+#include <assert.h> /* for assert */
+#include <stdio.h>  /* for scanf printf */
+#include <stdlib.h> /* for malloc, free */
+#include <string.h> /* for memcpy, memset */
+
+/**
+ * Function that combines two data to generate a new one
+ * The name of function might be misleading actually combine here signifies the
+ * fact that in segment trees we take partial result from two ranges and using
+ * paetial results we derive the result for joint range of those two ranges
+ * For Example: array(1,2,3,4,5,6) sum of range [0,2] = 6
+ * and sum of range [3,5] = 15 the combined sum of two range is 6+15=21
+ * Note: The function is same to binary function in Discrete Mathematics
+ * @param a pointer to first data
+ * @param b pointer to second data
+ * @param result pointer to memory location where result of combining a and b is
+ * to be stored
+ */
+typedef void (*combine_function)(const void *a, const void *b, void *result);
+
+/**
+ * @struct Segment tree
+ * This structures holds all the data that is required by a segment tree
+ */
+typedef struct segment_tree
+{
+    void *root;       /**< the root of formed segment tree */
+    void *identity;   /**< identity element for combine function */
+    size_t elem_size; /**< size of each data element */
+    size_t length;    /**< total size of array which segment tree represents*/
+    combine_function combine; /**< the function to be used to combine two node's
```

Comment:
if the description is longer, use the full- docblock syntax
```suggestion
    /** the function to be used to combine two node's
      * data to form parent's data 
      */
    combine_function combine; 
```